[WTF] Introduce validateKey to HashMap (2/3)

zakariaridouh opened the pull request:
#### df2b100eb9bbe0ffb024788c7652aa045359ed7e
<pre>
[WTF] Introduce validateKey to HashMap (2/3)
<a href="https://bugs.webkit.org/show_bug.cgi?id=279544">https://bugs.webkit.org/show_bug.cgi?id=279544</a>
<a href="https://rdar.apple.com/135826905">rdar://135826905</a>

Reviewed by Geoffrey Garen.

TLDR: `WTF::HashMap` is now safe by default. Use `UncheckedKeyHashMap`
to circumvent checks in performance sensitive code. The vast majority  of code should
use `HashMap`.

Elevate `ASSERT`s to `RELEASE_ASSERT`s in `WTF::HashTable::checkKey()` for `HashMap`, building on
a previous patch that a new type alias for an unsafe hashmap, aptly called `UncheckedKeyHashMap`.
Followup patches will expand on spreading this now safe HashMap across the codebase,
with the ultimate goal of having safe HashMap used in every non
performance sensitive portion of the code. [1]

[1] <a href="https://bugs.webkit.org/show_bug.cgi?id=278488">https://bugs.webkit.org/show_bug.cgi?id=278488</a>

* Source/WTF/wtf/Forward.h:
* Source/WTF/wtf/HashMap.h:
(WTF::validateKey&gt;::swap):
(WTF::validateKey&gt;::size const):
(WTF::validateKey&gt;::capacity const):
(WTF::validateKey&gt;::byteSize const):
(WTF::validateKey&gt;::isEmpty const):
(WTF::validateKey&gt;::begin):
(WTF::validateKey&gt;::end):
(WTF::validateKey&gt;::begin const):
(WTF::validateKey&gt;::end const):
(WTF::validateKey&gt;::find):
(WTF::validateKey&gt;::find const):
(WTF::validateKey&gt;::contains const):
(WTF::validateKey&gt;::get const):
(WTF::validateKey&gt;::inlineGet const):
(WTF::validateKey&gt;::remove):
(WTF::validateKey&gt;::inlineSet):
(WTF::validateKey&gt;::inlineAdd):
(WTF::validateKey&gt;::inlineEnsure):
(WTF::validateKey&gt;::set):
(WTF::validateKey&gt;::ensure):
(WTF::validateKey&gt;::add):
(WTF::validateKey&gt;::fastAdd):
(WTF::validateKey&gt;::getOptional const):
(WTF::validateKey&gt;::removeIf):
(WTF::validateKey&gt;::clear):
(WTF::validateKey&gt;::take):
(WTF::validateKey&gt;::takeOptional):
(WTF::validateKey&gt;::takeFirst):
(WTF::validateKey&gt;::checkConsistency const):
(WTF::validateKey&gt;::isValidKey):
(WTF::operator==):
(WTF::Y&gt;::swap): Deleted.
(WTF::Y&gt;::size const): Deleted.
(WTF::Y&gt;::capacity const): Deleted.
(WTF::Y&gt;::byteSize const): Deleted.
(WTF::Y&gt;::isEmpty const): Deleted.
(WTF::Y&gt;::begin): Deleted.
(WTF::Y&gt;::end): Deleted.
(WTF::Y&gt;::begin const): Deleted.
(WTF::Y&gt;::end const): Deleted.
(WTF::Y&gt;::find): Deleted.
(WTF::Y&gt;::find const): Deleted.
(WTF::Y&gt;::contains const): Deleted.
(WTF::Y&gt;::get const): Deleted.
(WTF::Y&gt;::inlineGet const): Deleted.
(WTF::Y&gt;::remove): Deleted.
(WTF::TableTraitsArg&gt;::inlineSet): Deleted.
(WTF::TableTraitsArg&gt;::inlineAdd): Deleted.
(WTF::TableTraitsArg&gt;::inlineEnsure): Deleted.
(WTF::TableTraitsArg&gt;::set): Deleted.
(WTF::TableTraitsArg&gt;::ensure): Deleted.
(WTF::TableTraitsArg&gt;::add): Deleted.
(WTF::TableTraitsArg&gt;::fastAdd): Deleted.
(WTF::Y&gt;::getOptional const): Deleted.
(WTF::Y&gt;::removeIf): Deleted.
(WTF::Y&gt;::clear): Deleted.
(WTF::Y&gt;::take): Deleted.
(WTF::Y&gt;::takeOptional): Deleted.
(WTF::Y&gt;::takeFirst): Deleted.
(WTF::Y&gt;::checkConsistency const): Deleted.
(WTF::Y&gt;::isValidKey): Deleted.
* Source/WTF/wtf/HashTable.h:
(WTF::validateKey&gt;::HashTable):
(WTF::validateKey&gt;::checkKey):
(WTF::validateKey&gt;::lookup):
(WTF::validateKey&gt;::inlineLookup):
(WTF::validateKey&gt;::lookupForReinsert):
(WTF::validateKey&gt;::fullLookupForWriting):
(WTF::validateKey&gt;::addUniqueForInitialization):
(WTF::validateKey&gt;::initializeBucket):
(WTF::validateKey&gt;::add):
(WTF::validateKey&gt;::addPassingHashCode):
(WTF::validateKey&gt;::reinsert):
(WTF::validateKey&gt;::find):
(WTF::validateKey&gt;::find const):
(WTF::validateKey&gt;::contains const):
(WTF::validateKey&gt;::removeAndInvalidateWithoutEntryConsistencyCheck):
(WTF::validateKey&gt;::removeAndInvalidate):
(WTF::validateKey&gt;::remove):
(WTF::validateKey&gt;::removeWithoutEntryConsistencyCheck):
(WTF::validateKey&gt;::removeIf):
(WTF::validateKey&gt;::allocateTable):
(WTF::validateKey&gt;::deallocateTable):
(WTF::validateKey&gt;::expand):
(WTF::validateKey&gt;::computeBestTableSize):
(WTF::validateKey&gt;::shrinkToBestSize):
(WTF::validateKey&gt;::deleteReleasedWeakBuckets):
(WTF::validateKey&gt;::rehash):
(WTF::validateKey&gt;::clear):
(WTF::validateKey&gt;::swap):
(WTF::=):
(WTF::validateKey&gt;::checkTableConsistency const):
(WTF::validateKey&gt;::checkTableConsistencyExceptSize const):
(WTF::KeyTraits&gt;::HashTable): Deleted.
(WTF::KeyTraits&gt;::checkKey): Deleted.
(WTF::KeyTraits&gt;::lookup): Deleted.
(WTF::KeyTraits&gt;::inlineLookup): Deleted.
(WTF::KeyTraits&gt;::lookupForReinsert): Deleted.
(WTF::KeyTraits&gt;::fullLookupForWriting): Deleted.
(WTF::KeyTraits&gt;::addUniqueForInitialization): Deleted.
(WTF::KeyTraits&gt;::initializeBucket): Deleted.
(WTF::KeyTraits&gt;::add): Deleted.
(WTF::KeyTraits&gt;::addPassingHashCode): Deleted.
(WTF::KeyTraits&gt;::reinsert): Deleted.
(WTF::KeyTraits&gt;::find): Deleted.
(WTF::KeyTraits&gt;::find const): Deleted.
(WTF::KeyTraits&gt;::contains const): Deleted.
(WTF::KeyTraits&gt;::removeAndInvalidateWithoutEntryConsistencyCheck): Deleted.
(WTF::KeyTraits&gt;::removeAndInvalidate): Deleted.
(WTF::KeyTraits&gt;::remove): Deleted.
(WTF::KeyTraits&gt;::removeWithoutEntryConsistencyCheck): Deleted.
(WTF::KeyTraits&gt;::removeIf): Deleted.
(WTF::KeyTraits&gt;::allocateTable): Deleted.
(WTF::KeyTraits&gt;::deallocateTable): Deleted.
(WTF::KeyTraits&gt;::expand): Deleted.
(WTF::KeyTraits&gt;::computeBestTableSize): Deleted.
(WTF::KeyTraits&gt;::shrinkToBestSize): Deleted.
(WTF::KeyTraits&gt;::deleteReleasedWeakBuckets): Deleted.
(WTF::KeyTraits&gt;::rehash): Deleted.
(WTF::KeyTraits&gt;::clear): Deleted.
(WTF::KeyTraits&gt;::swap): Deleted.
(WTF::KeyTraits&gt;::checkTableConsistency const): Deleted.
(WTF::KeyTraits&gt;::checkTableConsistencyExceptSize const): Deleted.

Canonical link: <a href="https://commits.webkit.org/285319@main">https://commits.webkit.org/285319@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8a056f958c6f33490b55fe8be647ee1d93a6ceb5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72229 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51650 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25017 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/76397 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/23438 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/74344 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/59454 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23260 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/56948 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15448 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75296 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46793 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62202 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/37381 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43454 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/19677 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21788 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/65358 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65355 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20037 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/78074 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/71483 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/16470 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/19192 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/65405 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/16517 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62226 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/64668 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15955 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12886 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6531 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/93264 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/47448 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/2232 "Found 1 new failure in dom/UserGestureIndicator.cpp") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/20528 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/48517 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49805 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/48260 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WTF/wtf/Forward.h:
```diff
@@ -133,9 +133,12 @@ struct IdentityExtractor;
 template<typename KeyTraits, typename MappedTraits> struct KeyValuePairTraits;
 template<typename KeyTypeArg, typename ValueTypeArg> struct KeyValuePair;
-template<typename Key, typename Value, typename Extractor, typename HashFunctions, typename Traits, typename KeyTraits> class HashTable;
+enum class ShouldValidateKey : bool { Yes, No };
```
This is reversed: `No` should come first, and then `Yes`. We want `No == 0 and Yes == 1`

zakariaridouh replied:
Fixed =)

cdumez commented on Source/WTF/wtf/HashTable.h:
```diff
@@ -631,44 +631,35 @@ DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(HashTable);
-    }
-
-#else // ASSERT_ENABLED
+        if constexpr (validateKey == ShouldValidateKey::No)
```
It's a little risky to not do this at call sites. If the function doesn't get inlined, then we'll pay a new function call in unsafe HashMaps, which could hurt perf.

rniwa commented on Source/WTF/wtf/HashTable.h:
```diff
@@ -631,44 +631,35 @@ DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(HashTable);
         if (!HashFunctions::safeToCompareToEmptyOrDeleted)
             return;
-        ASSERT(!HashTranslator::equal(KeyTraits::emptyValue(), key));
+        RELEASE_ASSERT(!HashTranslator::equal(KeyTraits::emptyValue(), key));
```
Wait, this would mean we no longer validate keys in debug builds for regular hash maps.
I don't think we want that behavior change.

zakariaridouh replied:
I believe 'RELEASE_ASSERT()' asserts on both Debug and Release builds, correct?

rniwa replied:
Yes but you're using ShouldValidateKey::No for regular hash maps so these release asserts would never run.

zakariaridouh replied:
Ah, got it! I'll add in a conditional guard that will force debug builds to validateKey and assert on invalid keys (similar to how behavior is before this change). Do you suggest a different approach?

mwyrzykowski commented on Source/WebGPU/WebGPU/RenderPassEncoder.h:
```diff
@@ -158,10 +158,10 @@ class RenderPassEncoder : public WGPURenderPassEncoderImpl, public RefCounted<Re
-    using EntryMap = HashMap<uint64_t, EntryUsage, DefaultHash<uint64_t>, WTF::UnsignedWithZeroKeyHashTraits<uint64_t>>;
-    HashMap<const void*, EntryMap> m_usagesForResource;
+    using EntryMap = SafeHashMap<uint64_t, EntryUsage, DefaultHash<uint64_t>, WTF::UnsignedWithZeroKeyHashTraits<uint64_t>>;
+    SafeHashMap<const void*, EntryMap> m_usagesForResource;
```
It is probably unnecessary to use `SafeHashMap` for `m_usagesForResource`, because we don't directly set anything coming from IPC and the key would need to be a pointer with value 2^64 - 1 to be unsafe, but also not really going to impact perf here

mwyrzykowski replied:
I have no objection to using SafeHashMap here however

cdumez replied:
Most HashMaps will be safe at some point so certainly no harm. That said, I agree it would be nice to prioritize HashMaps that are security sensitive.

geoffreygaren commented on Source/WTF/wtf/HashMap.h:
```diff
@@ -37,7 +37,7 @@ template<typename T> struct KeyValuePairKeyExtractor {
 };
 
-template<typename KeyArg, typename MappedArg, typename HashArg, typename KeyTraitsArg, typename MappedTraitsArg, typename TableTraitsArg>
+template<typename KeyArg, typename MappedArg, typename HashArg, typename KeyTraitsArg, typename MappedTraitsArg, typename TableTraitsArg, ShouldValidateKey validateKey>
```
I prefer a naming scheme where the variable matches the type: `ShouldValidateKey shouldValidateKey`.

It's more textually regular, if you're searching the project. It's usually the best name, since you gave the type a good name. And in this case, `validateKey` sounds like it might be a function you call to validate the key.

Minor point, but probably worth fixing.

geoffreygaren commented on Source/WTF/wtf/HashTable.h:
```diff
@@ -635,51 +635,42 @@ DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(HashTable);
             Traits::constructDeletedValue(deletedValue);
-            ASSERT(!HashTranslator::equal(Extractor::extract(deletedValue), key));
+            RELEASE_ASSERT(!HashTranslator::equal(Extractor::extract(deletedValue), key));
         }
```
If we're going to call this the safe HashMap, you need an else clause here:

```
} else {
    STATIC_ASSERT_NOT_REACHED_FOR_TYPE(Key, "ShouldValidateKey::Yes was specified, but no way to validate key.");
}
```

Eventually we should fill in how to validate keys where `!std::is_convertible_v<T, Key>)` and `! HashFunctions::safeToCompareToEmptyOrDeleted`. But a compile-time assertion is good enough to get started for now.

geoffreygaren reviewed (changes requested):
Looks like the right approach but I think we need two small fixes.

geoffreygaren reviewed (approved):
r=me

zakariaridouh committed:
[WTF] Introduce validateKey to HashMap (2/3)
https://bugs.webkit.org/show_bug.cgi?id=279544
rdar://135826905

Reviewed by Geoffrey Garen.

TLDR: `WTF::HashMap` is now safe by default. Use `UncheckedKeyHashMap`
to circumvent checks in performance sensitive code. The vast majority  of code should
use `HashMap`.

Elevate `ASSERT`s to `RELEASE_ASSERT`s in `WTF::HashTable::checkKey()` for `HashMap`, building on
a previous patch that a new type alias for an unsafe hashmap, aptly called `UncheckedKeyHashMap`.
Followup patches will expand on spreading this now safe HashMap across the codebase,
with the ultimate goal of having safe HashMap used in every non
performance sensitive portion of the code. [1]

[1] https://bugs.webkit.org/show_bug.cgi?id=278488

* Source/WTF/wtf/Forward.h:
* Source/WTF/wtf/HashMap.h:
(WTF::validateKey>::swap):
(WTF::validateKey>::size const):
(WTF::validateKey>::capacity const):
(WTF::validateKey>::byteSize const):
(WTF::validateKey>::isEmpty const):
(WTF::validateKey>::begin):
(WTF::validateKey>::end):
(WTF::validateKey>::begin const):
(WTF::validateKey>::end const):
(WTF::validateKey>::find):
(WTF::validateKey>::find const):
(WTF::validateKey>::contains const):
(WTF::validateKey>::get const):
(WTF::validateKey>::inlineGet const):
(WTF::validateKey>::remove):
(WTF::validateKey>::inlineSet):
(WTF::validateKey>::inlineAdd):
(WTF::validateKey>::inlineEnsure):
(WTF::validateKey>::set):
(WTF::validateKey>::ensure):
(WTF::validateKey>::add):
(WTF::validateKey>::fastAdd):
(WTF::validateKey>::getOptional const):
(WTF::validateKey>::removeIf):
(WTF::validateKey>::clear):
(WTF::validateKey>::take):
(WTF::validateKey>::takeOptional):
(WTF::validateKey>::takeFirst):
(WTF::validateKey>::checkConsistency const):
(WTF::validateKey>::isValidKey):
(WTF::operator==):
(WTF::Y>::swap): Deleted.
(WTF::Y>::size const): Deleted.
(WTF::Y>::capacity const): Deleted.
(WTF::Y>::byteSize const): Deleted.
(WTF::Y>::isEmpty const): Deleted.
(WTF::Y>::begin): Deleted.
(WTF::Y>::end): Deleted.
(WTF::Y>::begin const): Deleted.
(WTF::Y>::end const): Deleted.
(WTF::Y>::find): Deleted.
(WTF::Y>::find const): Deleted.
(WTF::Y>::contains const): Deleted.
(WTF::Y>::get const): Deleted.
(WTF::Y>::inlineGet const): Deleted.
(WTF::Y>::remove): Deleted.
(WTF::TableTraitsArg>::inlineSet): Deleted.
(WTF::TableTraitsArg>::inlineAdd): Deleted.
(WTF::TableTraitsArg>::inlineEnsure): Deleted.
(WTF::TableTraitsArg>::set): Deleted.
(WTF::TableTraitsArg>::ensure): Deleted.
(WTF::TableTraitsArg>::add): Deleted.
(WTF::TableTraitsArg>::fastAdd): Deleted.
(WTF::Y>::getOptional const): Deleted.
(WTF::Y>::removeIf): Deleted.
(WTF::Y>::clear): Deleted.
(WTF::Y>::take): Deleted.
(WTF::Y>::takeOptional): Deleted.
(WTF::Y>::takeFirst): Deleted.
(WTF::Y>::checkConsistency const): Deleted.
(WTF::Y>::isValidKey): Deleted.
* Source/WTF/wtf/HashTable.h:
(WTF::validateKey>::HashTable):
(WTF::validateKey>::checkKey):
(WTF::validateKey>::lookup):
(WTF::validateKey>::inlineLookup):
(WTF::validateKey>::lookupForReinsert):
(WTF::validateKey>::fullLookupForWriting):
(WTF::validateKey>::addUniqueForInitialization):
(WTF::validateKey>::initializeBucket):
(WTF::validateKey>::add):
(WTF::validateKey>::addPassingHashCode):
(WTF::validateKey>::reinsert):
(WTF::validateKey>::find):
(WTF::validateKey>::find const):
(WTF::validateKey>::contains const):
(WTF::validateKey>::removeAndInvalidateWithoutEntryConsistencyCheck):
(WTF::validateKey>::removeAndInvalidate):
(WTF::validateKey>::remove):
(WTF::validateKey>::removeWithoutEntryConsistencyCheck):
(WTF::validateKey>::removeIf):
(WTF::validateKey>::allocateTable):
(WTF::validateKey>::deallocateTable):
(WTF::validateKey>::expand):
(WTF::validateKey>::computeBestTableSize):
(WTF::validateKey>::shrinkToBestSize):
(WTF::validateKey>::deleteReleasedWeakBuckets):
(WTF::validateKey>::rehash):
(WTF::validateKey>::clear):
(WTF::validateKey>::swap):
(WTF::=):
(WTF::validateKey>::checkTableConsistency const):
(WTF::validateKey>::checkTableConsistencyExceptSize const):
(WTF::KeyTraits>::HashTable): Deleted.
(WTF::KeyTraits>::checkKey): Deleted.
(WTF::KeyTraits>::lookup): Deleted.
(WTF::KeyTraits>::inlineLookup): Deleted.
(WTF::KeyTraits>::lookupForReinsert): Deleted.
(WTF::KeyTraits>::fullLookupForWriting): Deleted.
(WTF::KeyTraits>::addUniqueForInitialization): Deleted.
(WTF::KeyTraits>::initializeBucket): Deleted.
(WTF::KeyTraits>::add): Deleted.
(WTF::KeyTraits>::addPassingHashCode): Deleted.
(WTF::KeyTraits>::reinsert): Deleted.
(WTF::KeyTraits>::find): Deleted.
(WTF::KeyTraits>::find const): Deleted.
(WTF::KeyTraits>::contains const): Deleted.
(WTF::KeyTraits>::removeAndInvalidateWithoutEntryConsistencyCheck): Deleted.
(WTF::KeyTraits>::removeAndInvalidate): Deleted.
(WTF::KeyTraits>::remove): Deleted.
(WTF::KeyTraits>::removeWithoutEntryConsistencyCheck): Deleted.
(WTF::KeyTraits>::removeIf): Deleted.
(WTF::KeyTraits>::allocateTable): Deleted.
(WTF::KeyTraits>::deallocateTable): Deleted.
(WTF::KeyTraits>::expand): Deleted.
(WTF::KeyTraits>::computeBestTableSize): Deleted.
(WTF::KeyTraits>::shrinkToBestSize): Deleted.
(WTF::KeyTraits>::deleteReleasedWeakBuckets): Deleted.
(WTF::KeyTraits>::rehash): Deleted.
(WTF::KeyTraits>::clear): Deleted.
(WTF::KeyTraits>::swap): Deleted.
(WTF::KeyTraits>::checkTableConsistency const): Deleted.
(WTF::KeyTraits>::checkTableConsistencyExceptSize const): Deleted.

Canonical link: https://commits.webkit.org/285319@main
